### PR TITLE
Review candidate routes

### PR DESCRIPTION
**Pull request by paulrobertlloyd**
## Context

Not only have we updated the ordering and labelling of items on a candidates application, but some of the paths we have refer to page titles we changed a long time ago, and no longer fully reflect their content or purpose.

## Changes proposed in this pull request

This is a proposal for changes we make to user facing paths. Controller names and file locations have not been changed, but should be if these recommendations are accepted.

Comments inline, but some high level notes:

* Replaces any `_` in paths with the GDS recommendation `-`
* Uses `delete` in user-facing paths, not `destroy`
* Uses `/thing/edit` not `/edit-thing`
* Tries to logically and consistently group related paths
* Removes a few redirects – do we still need these?

## Guidance to review

Agree with these recommendations?

## Link to Trello card

https://trello.com/c/lVDr7xYM/

## Things to check

- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] API release notes have been updated if necessary
- [ ] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training#azure-hosting-devops-pipeline)

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -48,76 +47,85 @@
...
+      get '/' => 'unsubmitted_application_form#show', as: :application_form
+      get '/review' => 'unsubmitted_application_form#review', as: :application_review
       get '/submit' => 'unsubmitted_application_form#submit_show', as: :application_submit_show
       post '/submit' => 'unsubmitted_application_form#submit', as: :application_submit
 
+      get '/submit-success' => 'submitted_application_form#submit_success', as: :application_submit_success
       get '/complete' => 'submitted_application_form#complete', as: :application_complete
       get '/review/submitted' => 'submitted_application_form#review_submitted', as: :application_review_submitted
-      get '/submit-success' => 'submitted_application_form#submit_success', as: :application_submit_success
-      get '/edit' => 'submitted_application_form#edit', as: :application_edit
```
Comment (paulrobertlloyd) [inline config/routes.rb:64]: This functionality no longer exists

File: config/routes.rb [hunk 2]
```diff
@@ -48,76 +47,85 @@
...
+      get '/review/submitted/:id' => 'application_form#review_previous_application', as: :review_previous_application
+
       get '/start-apply-again' => 'submitted_application_form#start_apply_again', as: :start_apply_again
       post '/apply-again' => 'submitted_application_form#apply_again', as: :apply_again
+
       get '/start-carry-over' => 'carry_over#start', as: :start_carry_over
       post '/carry-over' => 'carry_over#create', as: :carry_over
 
-      scope '/personal-details' do
+      scope '/personal-information' do
```
Comment (paulrobertlloyd) [inline config/routes.rb:70]: Changed to reflect new label for this section

File: config/routes.rb [hunk 3]
```diff
@@ -48,76 +47,85 @@
...
-        get '/nationalities' => 'personal_details/nationalities#new', as: :nationalities
-        patch '/nationalities' => 'personal_details/nationalities#create'
-        get '/nationalities/edit' => 'personal_details/nationalities#edit', as: :edit_nationalities
-        patch '/nationalities/edit' => 'personal_details/nationalities#update'
+
+        get '/nationality' => 'personal_details/nationalities#new', as: :nationalities
+        patch '/nationality' => 'personal_details/nationalities#create'
+        get '/nationality/edit' => 'personal_details/nationalities#edit', as: :edit_nationalities
+        patch '/nationality/edit' => 'personal_details/nationalities#update'
+
```
Comment (paulrobertlloyd) [inline config/routes.rb:80]: The page title is ‘What is your nationality?’ and the review label is ‘nationality’.

File: config/routes.rb [hunk 4]
```diff
@@ -48,76 +47,85 @@
...
+        get '/review' => 'personal_statement/subject_knowledge#show', as: :subject_knowledge_show
+        patch '/complete' => 'personal_statement/subject_knowledge#complete', as: :subject_knowledge_complete
+      end
+
+      scope '/interview-needs' do
+        get '/' => 'personal_statement/interview_preferences#edit', as: :edit_interview_preferences
+        patch '/' => 'personal_statement/interview_preferences#update'
+        get '/review' => 'personal_statement/interview_preferences#show', as: :interview_preferences_show
+        patch '/complete' => 'personal_statement/interview_preferences#complete', as: :interview_preferences_complete
+      end
```
Comment (paulrobertlloyd) [inline config/routes.rb:114]: Move `interview-preferences` from outside `personal-statement` - no other section has this sort of folder structure, and also this section no longer lives under the ‘Personal statement’ heading in the application.

Also, rename it `interview-needs` to reflect the section title.

File: config/routes.rb [hunk 5]
```diff
@@ -48,76 +47,85 @@
...
+        patch '/complete' => 'personal_statement/becoming_a_teacher#complete', as: :becoming_a_teacher_complete
       end
 
-      scope '/training-with-a-disability' do
+      scope '/subject-knowledge' do
+        get '/' => 'personal_statement/subject_knowledge#edit', as: :edit_subject_knowledge
+        patch '/' => 'personal_statement/subject_knowledge#update'
+        get '/review' => 'personal_statement/subject_knowledge#show', as: :subject_knowledge_show
+        patch '/complete' => 'personal_statement/subject_knowledge#complete', as: :subject_knowledge_complete
+      end
```
Comment (paulrobertlloyd) [inline config/routes.rb:107]: Move `subject-knowledge` from outside `personal-statement` - no other section has this sort of folder structure.

File: config/routes.rb [hunk 6]
```diff
@@ -48,76 +47,85 @@
...
-
-        get '/interview-preferences' => 'personal_statement/interview_preferences#edit', as: :edit_interview_preferences
-        patch '/interview-preferences' => 'personal_statement/interview_preferences#update'
-        get '/interview-preferences/review' => 'personal_statement/interview_preferences#show', as: :interview_preferences_show
-        patch '/interview-preferences/complete' => 'personal_statement/interview_preferences#complete', as: :interview_preferences_complete
+        get '/' => 'personal_statement/becoming_a_teacher#edit', as: :edit_becoming_a_teacher
+        patch '/' => 'personal_statement/becoming_a_teacher#update'
+        get '/review' => 'personal_statement/becoming_a_teacher#show', as: :becoming_a_teacher_show
+        patch '/complete' => 'personal_statement/becoming_a_teacher#complete', as: :becoming_a_teacher_complete
       end
```
Comment (paulrobertlloyd) [inline config/routes.rb:100]: Removing the sections previously nested below `personal-statement` means we can now use this for what was previously `/personal-statement/becoming-a-teacher`

File: config/routes.rb [hunk 7]
```diff
@@ -48,76 +47,85 @@
...
+
+      scope '/additional-support' do
         get '/' => 'training_with_a_disability#edit', as: :edit_training_with_a_disability
         patch '/' => 'training_with_a_disability#update'
         get '/review' => 'training_with_a_disability#show', as: :training_with_a_disability_show
         patch '/complete' => 'training_with_a_disability#complete', as: :training_with_a_disability_complete
       end
 
-      scope '/contact-details' do
+      scope '/contact-information' do
```
Comment (paulrobertlloyd) [inline config/routes.rb:123]: Update to match new section title.

File: config/routes.rb [hunk 8]
```diff
@@ -48,76 +47,85 @@
...
         get '/' => 'contact_details/base#edit', as: :contact_details_edit_base
         patch '/' => 'contact_details/base#update'
 
-        get '/address_type' => 'contact_details/address_type#edit', as: :contact_details_edit_address_type
-        patch '/address_type' => 'contact_details/address_type#update'
+        get '/address-type' => 'contact_details/address_type#edit', as: :contact_details_edit_address_type
+        patch '/address-type' => 'contact_details/address_type#update'
```
Comment (paulrobertlloyd) [inline config/routes.rb:128]: Use hyphens in paths, not underscores.

File: config/routes.rb [hunk 9]
```diff
@@ -144,8 +152,8 @@
         get '/country' => 'gcse/institution_country#edit', as: :gcse_details_edit_institution_country
         patch '/country' => 'gcse/institution_country#update'
 
-        get '/naric-reference' => 'gcse/naric_reference#edit', as: :gcse_details_edit_naric_reference
-        patch '/naric-reference' => 'gcse/naric_reference#update'
+        get '/naric' => 'gcse/naric_reference#edit', as: :gcse_details_edit_naric_reference
+        patch '/naric' => 'gcse/naric_reference#update'
```
Comment (paulrobertlloyd) [inline config/routes.rb:156]: Use `naric` consistently in qualification sections.

File: config/routes.rb [hunk 10]
```diff
@@ -184,7 +192,7 @@
         delete '/delete/:id' => 'work_history/destroy#destroy'
       end
 
-      scope '/school-experience' do
+      scope '/unpaid-experience' do
```
Comment (paulrobertlloyd) [inline config/routes.rb:195]: Update to reflect section title.

File: config/routes.rb [hunk 11]
```diff
@@ -217,15 +225,15 @@
...
         patch '/:id/institution/edit' => 'degrees/institution#update'
 
-        get '/:id/completion_status' => 'degrees/completion_status#new', as: :degree_completion_status
-        post '/:id/completion_status' => 'degrees/completion_status#create'
-        get '/:id/completion_status/edit' => 'degrees/completion_status#edit', as: :edit_degree_completion_status
-        patch '/:id/completion_status/edit' => 'degrees/completion_status#update'
+        get '/:id/completion-status' => 'degrees/completion_status#new', as: :degree_completion_status
+        post '/:id/completion-status' => 'degrees/completion_status#create'
+        get '/:id/completion-status/edit' => 'degrees/completion_status#edit', as: :edit_degree_completion_status
+        patch '/:id/completion-status/edit' => 'degrees/completion_status#update'
```
Comment (paulrobertlloyd) [inline config/routes.rb:231]: Use hyphens, not underscores.

File: config/routes.rb [hunk 12]
```diff
@@ -204,8 +212,8 @@
       scope '/degrees' do
         get '/' => 'degrees/type#new', as: :new_degree
         post '/' => 'degrees/type#create'
-        get '/:id/type/edit' => 'degrees/type#edit', as: :edit_degree_type
-        patch '/:id/type/edit' => 'degrees/type#update'
+        get '/:id/edit' => 'degrees/type#edit', as: :edit_degree_type
+        patch '/:id/edit' => 'degrees/type#update'
```
Comment (paulrobertlloyd) [inline config/routes.rb:216]: Be consistent in how we address editing the first question in a section.

File: config/routes.rb [hunk 13]
```diff
@@ -217,15 +225,15 @@
...
+        patch '/:id/completion-status/edit' => 'degrees/completion_status#update'
 
-        get '/:id/naric-statement' => 'degrees/naric_statement#new', as: :degree_naric_statement
-        post '/:id/naric-statement' => 'degrees/naric_statement#create'
-        get '/:id/naric-statement/edit' => 'degrees/naric_statement#edit', as: :edit_degree_naric_statement
-        patch '/:id/naric-statement/edit' => 'degrees/naric_statement#update'
+        get '/:id/naric' => 'degrees/naric_statement#new', as: :degree_naric_statement
+        post '/:id/naric' => 'degrees/naric_statement#create'
+        get '/:id/naric/edit' => 'degrees/naric_statement#edit', as: :edit_degree_naric_statement
+        patch '/:id/naric/edit' => 'degrees/naric_statement#update'
```
Comment (paulrobertlloyd) [inline config/routes.rb:236]: Use `naric` consistently in qualification sections

File: config/routes.rb [hunk 14]
```diff
@@ -254,7 +262,6 @@
         get '/choose' => 'course_choices/have_you_chosen#ask', as: :course_choices_choose
         post '/choose' => 'course_choices/have_you_chosen#decide'
         get '/find-a-course' => 'course_choices/have_you_chosen#go_to_find', as: :go_to_find
-        get '/find_a_course', to: redirect('/candidate/application/courses/find-a-course')
```
Comment (paulrobertlloyd) [inline config/routes.rb:257]: Do we still need this redirect?

File: config/routes.rb [hunk 15]
```diff
@@ -272,9 +279,7 @@
         get '/apply-on-ucas/provider/:provider_id/course/:course_id' => 'course_choices/ucas#with_course', as: :course_choices_ucas_with_course
 
         get '/confirm-selection/:course_id' => 'find_course_selections#confirm_selection', as: :course_confirm_selection
-        get '/confirm_selection/:course_id', to: redirect('/candidate/application/courses/confirm-selection/%{course_id}')
         post '/complete-selection/:course_id' => 'find_course_selections#complete_selection', as: :course_complete_selection
-        get '/complete_selection/:course_id', to: redirect('/candidate/application/courses/complete-selection/%{course_id}')
```
Comment (paulrobertlloyd) [inline config/routes.rb:277]: Do we still need this 2 redirects?

File: config/routes.rb [hunk 16]
```diff
@@ -317,11 +320,13 @@
       scope '/english-as-a-foreign-language' do
         get '/' => 'english_foreign_language/start#new', as: :english_foreign_language_start
         post '/' => 'english_foreign_language/start#create'
-        get '/proficiency/edit' => 'english_foreign_language/start#edit', as: :english_foreign_language_edit_start
-        patch '/proficiency/edit' => 'english_foreign_language/start#update'
+        get '/edit' => 'english_foreign_language/start#edit', as: :english_foreign_language_edit_start
+        patch '/edit' => 'english_foreign_language/start#update'
```
Comment (paulrobertlloyd) [inline config/routes.rb:324]: Be consistent in how we address editing the first question in a section.

File: config/routes.rb [hunk 17]
```diff
@@ -317,11 +320,13 @@
...
 
         get '/type' => 'english_foreign_language/type#new', as: :english_foreign_language_type
         post '/type' => 'english_foreign_language/type#create'
+        # get '/type/edit' ?
+        # patch '/type/edit' ?
```
Comment (paulrobertlloyd) [inline config/routes.rb:329]: Are these 2 paths missing for a reason?

File: config/routes.rb [hunk 18]
```diff
@@ -297,15 +302,13 @@
...
-
-        get '/new' => 'other_qualifications/details#new', as: :other_qualification_details
-        patch '/new' => 'other_qualifications/details#create'
-
-        get '/edit-details/:id' => 'other_qualifications/details#edit', as: :edit_other_qualification_details
-        patch '/edit-details/:id' => 'other_qualifications/details#update'
+        get '/details' => 'other_qualifications/details#new', as: :other_qualification_details
+        patch '/details' => 'other_qualifications/details#create'
+        get '/details/edit/:id' => 'other_qualifications/details#edit', as: :edit_other_qualification_details
+        patch '/details/edit/:id' => 'other_qualifications/details#update'
```
Comment (paulrobertlloyd) [inline config/routes.rb:311]: Use `/thing/edit` not `/edit-thing`

File: config/routes.rb [hunk 19]
```diff
@@ -347,66 +352,68 @@
...
 
         get '/review' => 'references/review#show', as: :references_review
-        get '/review/destroy-referee/:id' => 'references/review#confirm_destroy_referee', as: :destroy_referee
-        get '/review/destroy-reference/:id' => 'references/review#confirm_destroy_reference', as: :destroy_reference
-        get '/review/destroy-reference-request/:id' => 'references/review#confirm_destroy_reference_request', as: :delete_reference_request
-        delete '/review/destroy-reference/:id' => 'references/review#destroy'
+        get '/review/delete-referee/:id' => 'references/review#confirm_destroy_referee', as: :destroy_referee
+        get '/review/delete-reference/:id' => 'references/review#confirm_destroy_reference', as: :destroy_reference
+        get '/review/delete-request/:id' => 'references/review#confirm_destroy_reference_request', as: :delete_reference_request
+        delete '/review/delete-reference/:id' => 'references/review#destroy'
```
Comment (paulrobertlloyd) [inline config/routes.rb:380]: Use `delete` in user facing paths, not `destroy`

File: config/routes.rb [hunk 20]
```diff
@@ -347,66 +352,68 @@
...
         get 'review/cancel/:id' => 'references/review#confirm_cancel', as: :confirm_cancel_reference
         patch 'review/cancel/:id' => 'references/review#cancel', as: :cancel_reference
 
         get '/request/:id' => 'references/request#new', as: :references_new_request
         post '/request/:id' => 'references/request#create', as: :references_create_request
 
-        get '/retry_request/:id' => 'references/retry_request#new', as: :references_retry_request
-        post '/retry_request/:id' => 'references/retry_request#create'
+        get '/retry-request/:id' => 'references/retry_request#new', as: :references_retry_request
+        post '/retry-request/:id' => 'references/retry_request#create'
```
Comment (paulrobertlloyd) [inline config/routes.rb:389]: Use hyphens, not underscores.

File: config/routes.rb [hunk 21]
```diff
@@ -347,66 +352,68 @@
...
 
         get '/reminder/:id' => 'references/reminder#new', as: :references_new_reminder
         post '/reminder/:id' => 'references/reminder#create'
 
-        get '/candidate_name/:id' => 'references/candidate_name#new', as: :references_new_candidate_name
-        post '/candidate_name/:id' => 'references/candidate_name#create', as: :references_create_candidate_name
+        get '/candidate-name/:id' => 'references/candidate_name#new', as: :references_new_candidate_name
+        post '/candidate-name/:id' => 'references/candidate_name#create', as: :references_create_candidate_name
```
Comment (paulrobertlloyd) [inline config/routes.rb:395]: Use hyphens, not underscores.

**Review by paulrobertlloyd — COMMENTED** [review]
Comments inline about the various changes/recommendations.

**Comment by paulrobertlloyd** [comment]
Testing this locally, links in the application menu no longer work, even though they should change based if using path helpers. Are the paths hard coded somewhere @TobyRet?

**Comment by paulrobertlloyd** [comment]
If we make these changes, question is do we add redirects, and if so, to what extent? Maybe it’s safe to assume that these URLs are not bookmarked or saved somewhere (although browsers will have them in history).

@frankieroberto made the point that this could cause issues if the URLs change tries to save an answer as this change is delayed – though maybe that’d still be the case even with redirects?

Maybe add redirects, but only maintain them for a limited period?

File: config/routes.rb [hunk 22]
```diff
@@ -297,15 +302,13 @@
       scope '/other-qualifications' do
         get '/' => 'other_qualifications/type#new', as: :other_qualification_type
         post '/' => 'other_qualifications/type#create'
+        get '/edit/:id' => 'other_qualifications/type#edit', as: :edit_other_qualification_type
+        patch '/edit/:id' => 'other_qualifications/type#update'
```
Comment (frankieroberto) [inline config/routes.rb:306]: I think `/edit` should always be the last bit of the URL?

```suggestion
        get '/:id/edit' => 'other_qualifications/type#edit', as: :edit_other_qualification_type
        patch '/:id/edit' => 'other_qualifications/type#update'
```

**Review by frankieroberto — COMMENTED** [review]
All makes sense to me.

Agree we should avoid nesting URLs where possible - particularly as then it makes it easier to group and re-group sections without affecting the URLs.

I think we should use `/thing/edit` rather than `/edit/thing` – mainly to follow the convention that if you remove bits of the URL from the right up to the previous slash, then you get to an actual page rather than a 404.

For this reason, I have in the past sometimes used the `/thing` route for the "review" page, rather than `/thing/review` – but I think in our service most of those `/thing` pages are used for routing or guard questions?

For deletion confirmation pages, perhaps these should be at the end of the URL too - ie `/thing/delete` rather than `/delete-thing` or `/delete/thing`?

Personally, I prefer `_` to `-` in URLs, as I think it can make them easier to read, the underscores more closely resembling spaces. Plus then hyphens can be used for hyphenated words only, so that `/man-eating-shark` can be disambiguated into `/man-eating_shark`. 😄  But I'll accept this is a minority view, and underscores can be harder for people to type on a keyboard, so I'll let this go 🙊

**Comment by paulrobertlloyd** [comment]
> Personally, I prefer _ to - in URLs, as I think it can make them easier to read, the underscores more closely resembling spaces.

I personally use underscores too - mainly as a small form of protest at Google who popularised their usage ✊🏻 But the GDS recommendation is hyphens, and we use hyphens everywhere else, so we should stick with that.

File: config/routes.rb (hunk 17 above)
Comment (TobyRet) [reply config/routes.rb:329]: mmm interesting - the controller does not appear to have an edit or update action. 
https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/app/controllers/candidate_interface/english_foreign_language/type_controller.rb

@davidgisbey or @malcolmbaig can you shed some light on this?

**Comment by TobyRet** [comment]
> Testing this locally, links in the application menu no longer work, even though they should have updated if using path helpers. Are the paths hard coded somewhere @TobyRet?

@paulrobertlloyd links in the application menu work for me. Did you restart your server?

**Comment by paulrobertlloyd** [comment]
> Did you restart your server?

Ah, that’d be it! 🤦🏼‍♂️

File: config/routes.rb (hunk 22 above)
Comment (paulrobertlloyd) [reply config/routes.rb:306]: So always `/[section-name]/:id/edit` or `/[section-name]/:id/edit/[question]`? That would effect all sections in the candidate interface, which might be a consideration if we were to change this.

Comment (frankieroberto) [reply config/routes.rb:306]: @paulrobertlloyd If it's one-page per question, I think you could do `/[section-name]/:id/[question]` – no need for the word "edit" at all?

File: config/routes.rb [hunk 23]
```diff
@@ -68,16 +68,16 @@
...
         patch '/edit' => 'personal_details/base#update'
 
-        get '/nationalities' => 'personal_details/nationalities#new', as: :nationalities
-        patch '/nationalities' => 'personal_details/nationalities#create'
-        get '/nationalities/edit' => 'personal_details/nationalities#edit', as: :edit_nationalities
-        patch '/nationalities/edit' => 'personal_details/nationalities#update'
+        get '/nationality' => 'personal_details/nationalities#new', as: :nationalities
+        patch '/nationality' => 'personal_details/nationalities#create'
+        get '/nationality/edit' => 'personal_details/nationalities#edit', as: :edit_nationalities
+        patch '/nationality/edit' => 'personal_details/nationalities#update'
```
Comment (paulrobertlloyd) [inline config/routes.rb:80]: Use `nationality` as this is the word we use in the question title and on the review page.

File: config/routes.rb (hunk 17 above)
Comment (malcolmbaig) [reply config/routes.rb:333]: The TypeController doesn't create a resource, it just handles POSTs from the type form and decides which EFL controller to send you to next. Therefore, there isn't really a need for edit/update actions.

When a candidate chooses to "edit" their type, nothing actually happens to their saved record until they submit the subsequent EFL qualification form.

**Pull request closed without merging**